### DISCUSSION TITLE
STM32F4-Discovery

### DISCUSSION THREAD

**Discussion by ekawahyu** [xpack-dev-tools/qemu-arm-xpack#20] [Q&A]
With qemu-system-gnuarmeclipse being deprecated, is there a guide (or hints?) on how to emulate STM32F4-Discovery with emu-system-arm?

**Comment by ilg-ul** [comment]
Unfortunately not. I updated my tests to use one of the boards supported by the upstream QEMU. 

Maybe late this year to have some time for more work on QEMU, including to migrate STM32F4-Discovery and the other existing boards.

**Reply by ekawahyu** [reply]
Thank you for your response @ilg-ul. I really wanted to help, but I have a little to no knowledge about QEMU. If you could give an example of one platform (ARM Cortex-M) that has been successfully ported from qemu-system-gnuarmeclipse to qemu-system-arm, I can try to look at the changes required and may be able to help porting the STM32F4 Discovery.

**Reply by ilg-ul** [reply]
Thank you for your offer to help.

qemu-system-gnuarmeclipse has a troubled history, it was created to match the requirements of Eclipse, but the upstream maintainers were not interested in any of the new features. Actually interest for Cortex-M was very low.

Recently I saw that support for Cortex-M improved, but it is still not possible to integrate it with Eclipse. However it is usable for CI tests, but with some limitations, for example no M0/M0+.

I'll try to discuss with the upstream maintainers and search for ways to improve things, with the goal to add more boards like STM32F4 Discovery, but it won't be easy, and I don't have the resources to actively maintain a fork with many changes.

**Reply by ekawahyu** [reply]
Sure, I cannot promise anything either before looking at the code base and have at least some understanding on how QEMU works under the hood. But yeah, please do let me know when there is anything I can help with Cortex-M in general, or STM32 in particular.

**Reply by ilg-ul** [reply]
I suggest you take a look at the current source code. It is very focused on running Linux kernels on larger machines, and Cortex-M is supported accidentally, but you need to understand how it works first.

In qemu-system-gnuarmeclipse I took a slightly different path, I created a set of generic objects implementing peripheral/register/bitfield, which, combined, were able to better define microcontroller registers. The interesting part was that these definitions were generated automatically from the SVD files, and only the functionality required on read/write needed to be implemented manually.

For a future qemu I plan do take this to a higher level, and generate code that includes user guards, like CubeMX does, where the read-write callbacks implemented by the user will reside, so the code can be regenerated and the user code be preserved.

But it is a long way to such a solution...

**Reply by ekawahyu** [reply]
First question, will Multipass work well for QEMU development? Or should I do development on native Linux installation?

**Reply by ilg-ul** [reply]
I have no experience with Multipass, so perhaps my question is stupid, but what is the expected relationship between Multipass, Cortex-M and QEMU?

QEMU is an emulator that, in the Cortex-M case, does a kind of just in time compilation of thumb instructions into native instructions, and runs them.

**Reply by ekawahyu** [reply]
I guess the stupid question was on me ha..ha.. because I didn't realize that I was actually trying to run QEMU in Multipass. Anyway, it will save the trouble to develop QEMU natively. I am on Apple M1 by the way, but I also have access to amd64 machine. What would be the best machine to start developing QEMU in your opinion? I mean to get started quickly with less hiccup.

**Reply by ilg-ul** [reply]
Since I already did a binary xPack with QEMU for Apple Silicon, it should be ok on your M1 system too.

I do have some details in the README-DEVELOP page on how to build natively and run debug sessions, but the instructions refer to Visual Studio Code, which is not yet fully available on M1, so you might try to use Xcode.

**Reply by ilg-ul** [reply]
To answer your question, the best machine to develop QEMU would probably be a Linux box.

**Reply by LAPSyLAB** [reply]
We're using STM32F4 emulation for teaching assembly programming at the university, so we would kindly ask that at least qemu-system-gnuarmeclipse binary stays in the distribution, until complete migration happens... 

Thanks, regards.

**Reply by ilg-ul** [reply]
I'll try to keep it for a while, but I advise on using qemu-system-arm, and not counting on the migration to happen very soon.
